mailpoet/lib/Form/DisplayFormInWPContent.php

```diff
@@ -113,7 +125,7 @@ private function shouldDisplay(): bool {
     // this code ensures that we display the form only on a page which is related to single post
     if (!$this->wp->isSingle() && !$this->wp->isPage()) $result = $this->wp->applyFilters('mailpoet_display_form_is_single', false);
 
-    if ($this->displayFormInProductListPage()) $result = true;
+    if ($this->inWooProductLoop) $result = $this->displayFormInProductListPage();
```

Comment:
I need help understanding this change. Please can you explain it? Also, a comment in the code can be helpful because this logic is very complex.

Reply:
This method `displayFormInProductListPage()`, checks if we can display forms on Woo Shop page.

The bug we had previously: This function `shouldDisplay()` was being called within the `woocommerce_product_loop_end` filter and was always returning true (because that was the default value) instead of false or true depending on the result of the `displayFormInProductListPage()` method check.

For example

When you have the [products] shortcode in multiple locations on the page. This filter `woocommerce_product_loop_end` would be called at the end of each of those product loops. 

if `shouldDisplay()` returns true for each of those product loops, the form would be displayed multiple times.

The update here ensures the form should only display depending on the result of the `displayFormInProductListPage()` method check (if we are within the `woocommerce_product_loop_end` filter).